### PR TITLE
Add Tuya Number support

### PR DESCRIPTION
**Pull request by frankiboy1**
# What does this implement/fix? 

This PR adds support for Tuya Number. Tuya devices have currently only support for binary_sensor, climate, cover, fan, light, sensor and switch. Without number support, it's not possible to support generic devices without trying to map them into one of the existing types. With number support, it should be possible to support any tuya device using the generic types binary_sensor, sensor, switch and number.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1634

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [X] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
  - platform: "tuya"
    name: "MyNumber"
    number_datapoint: 102
    min_value: 0
    max_value: 17
    step: 1
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by oxan** [comment]
See also #2311.

**Comment by frankiboy1** [comment]
> See also #2311.

I wasn't aware of this PR. I had a quick look at that PR and it's mostly doing the same as mine. My version supports both integer and enum datapoints, which should be a requirement.

**Comment by frankiboy1** [comment]
This PR should close this feature request:

https://github.com/esphome/feature-requests/issues/1328

Not sure how to link the feature request to the pull request, or if this happens automatically due to this mention.

**Comment by frankiboy1** [comment]
Not sure what happened here. I tried to rebase my changes and somehow it ended up as part of the PR. I'll try to clean up after me..

**Comment by frankiboy1** [comment]
I have reverted the rebase I recently made, so I think everything should be correct now. I can see that some labels were not removed, and some assignees/reviewers are still on the list. Sorry for screwing this up! ;(

File: esphome/components/tuya/number/__init__.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+DEPENDENCIES = ["tuya"]
+CODEOWNERS = ["@frankiboy1"]
+
+TuyaNumber = tuya_ns.class_("TuyaNumber", number.Number, cg.Component)
+
+
+def validate_min_max(config):
+    if config[CONF_MAX_VALUE] <= config[CONF_MIN_VALUE]:
+        raise cv.Invalid("max_value must be greater than min_value")
+    return config
```
Comment (jesserockz) [inline esphome/components/tuya/number/__init__.py:22]: You are not using this validation

**Review by jesserockz — CHANGES_REQUESTED** [review]
Just one thing

File: esphome/components/tuya/number/__init__.py (hunk 1 above)
Comment (frankiboy1) [reply esphome/components/tuya/number/__init__.py:22]: Should be fixed with the new commit.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**